oxide-vpc/tests/integration_tests.rs

```diff
@@ -1602,6 +1617,238 @@ fn guest_to_internet() {
     }
 }
 
+// Verify that an ICMP Echo request has its identifier rewritten by
+// SNAT.
+#[test]
+fn snat_icmp4_echo_rewrite() {
+    let g1_cfg = g1_cfg();
+    let v2p = Arc::new(Virt2Phys::new());
+    let mut ameta = ActionMeta::new();
+    let mut g1 = oxide_net_setup("g1_port", &g1_cfg, v2p.clone());
+    g1.port.start();
+    set_state!(g1, PortState::Running);
+    let dst_ip: Ipv4Addr = "45.55.45.205".parse().unwrap();
+    let ident = 7;
+    let mut seq_no = 777;
+    let data = b"reunion\0";
+
+    // Add router entry that allows g1 to route to internet.
+    router::add_entry(
+        &g1.port,
+        IpCidr::Ip4("0.0.0.0/0".parse().unwrap()),
+        RouterTarget::InternetGateway,
+    )
+    .unwrap();
+    incr!(g1, ["epoch", "router.rules_out"]);
+    let mapped_port = g1_cfg.snat().ports.clone().rev().next().unwrap();
+
+    // ================================================================
+    // Verify echo request rewrite.
+    // ================================================================
+    let mut pkt1 = gen_icmp_echo_req(
+        g1_cfg.private_mac,
+        g1_cfg.gateway_mac,
+        g1_cfg.ipv4().private_ip.into(),
+        dst_ip.into(),
+        ident,
+        seq_no,
+        &data[..],
+    );
+
+    let res = g1.port.process(Out, &mut pkt1, &mut ameta);
+    assert!(matches!(res, Ok(Modified)), "bad result: {:?}", res);
+    incr!(
+        g1,
+        [
+            "fw.flows_out",
+            "fw.flows_in",
+            "nat.flows_out",
+            "nat.flows_in",
+            "uft.flows_out"
+        ]
+    );
+
+    // Ether + IPv6 + UDP + Geneve + Ether + IPv4
+    assert_eq!(pkt1.body_offset(), 14 + 40 + 8 + 8 + 14 + 20);
```

Comment:
Nit: Might want to use the `const` sizes we define for these?

Reply:
Great point. I started making const values for these. While doing this I also moved some of the `<TYPE>_HDR_SZ` constants into their respective header types: e.g. `EtherHdr::SIZE`. This lead to more refactoring of other files but I do feel like it's more idiomatic rust code so just did it while I was here.